src/red_black_tree.rs

```diff
@@ -933,6 +959,93 @@ impl<
     }
 }
 
+pub struct RedBlackTreeDrainFilter<
+    'a,
+    K: Debug + PartialOrd + Ord + Copy + Clone + Default + Pod + Zeroable,
+    V: Default + Copy + Clone + Pod + Zeroable,
+    P: for<'p> Fn(&'p K, &'p V) -> bool,
+    const MAX_SIZE: usize,
+> {
+    tree: &'a mut RedBlackTree<K, V, MAX_SIZE>,
+    fwd_stack: Vec<u32>,
+    fwd_ptr: u32,
+    fwd_node: Option<u32>,
+    _rev_stack: Vec<u32>,
+    _rev_ptr: u32,
+    rev_node: Option<u32>,
+    terminated: bool,
+    /// Keeps addr's of nodes with predicate = true, to remove upon dropping.
+    /// It is possible to instead drop during iteration and update fwd_ptr & fwd_stack
+    remove: Vec<u32>,
+    predicate: P,
+}
+
+impl<
+        'a,
+        K: Debug + PartialOrd + Ord + Copy + Clone + Default + Pod + Zeroable,
+        V: Default + Copy + Clone + Pod + Zeroable,
+        P: for<'p> Fn(&'p K, &'p V) -> bool,
+        const MAX_SIZE: usize,
+    > Drop for RedBlackTreeDrainFilter<'a, K, V, P, MAX_SIZE>
+{
+    fn drop(&mut self) {
+        for node_index in core::mem::take(&mut self.remove) {
+            self.tree._remove_tree_node(node_index);
+        }
+    }
+}
+
+impl<
+        'a,
+        K: Debug + PartialOrd + Ord + Copy + Clone + Default + Pod + Zeroable,
+        V: Default + Copy + Clone + Pod + Zeroable,
+        P: for<'p> Fn(&'p K, &'p V) -> bool,
+        const MAX_SIZE: usize,
+    > Iterator for RedBlackTreeDrainFilter<'a, K, V, P, MAX_SIZE>
+{
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while !self.terminated && (!self.fwd_stack.is_empty() || self.fwd_ptr != SENTINEL) {
+            if self.fwd_ptr != SENTINEL {
+                self.fwd_stack.push(self.fwd_ptr);
+                self.fwd_ptr = self.tree.get_left(self.fwd_ptr);
+            } else {
+                let current_node = self.fwd_stack.pop();
+                if current_node == self.rev_node {
+                    self.terminated = true;
+                    return None;
+                }
+                self.fwd_node = current_node;
+                let ptr = self.fwd_node.unwrap();
+
+                // Get node, check predicate.
+                // If predicate, remove and return
+                let node = self
+                    .tree
+                    .allocator
+                    .nodes
+                    .get((ptr - 1) as usize)
+                    .unwrap()
+                    .get_value();
```

Comment:
is there not a helper function to perform this lookup?

Reply:
This part was mostly copied from the iterator implementation, but perhaps

Reply:
switched to `get_node`. much neater